core/src/services/upyun/writer.rs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use http::StatusCode;
+
+use crate::raw::*;
+use crate::*;
+
+use super::core::{parse_initiate_part, UpYunCore};
+use super::error::parse_error;
+
+pub type UpYunWriters = oio::MultipartUploadWriter<UpYunWriter>;
+
+pub struct UpYunWriter {
+    core: Arc<UpYunCore>,
+    op: OpWrite,
+    path: String,
+}
+
+impl UpYunWriter {
+    pub fn new(core: Arc<UpYunCore>, op: OpWrite, path: String) -> Self {
+        UpYunWriter { core, op, path }
+    }
+}
+
+#[async_trait]
+impl oio::MultipartUploadWrite for UpYunWriter {
+    async fn write_once(&self, size: u64, body: AsyncBody) -> Result<()> {
+        let mut req = self
+            .core
+            .upload(&self.path, Some(size), &self.op, body)
+            .await?;
+
+        self.core.sign(&mut req).await?;
```

Comment:
Could this line be merged into upload fn?

Reply:
fix